lib/twilio/device.ts

```diff
@@ -556,62 +561,68 @@ class Device extends EventEmitter {
+          customParameters = connectTokenParts.customParameters;
+          parameters = connectTokenParts.parameters;
+          signalingReconnectToken = connectTokenParts.signalingReconnectToken;
+        } catch {
+          throw new InvalidArgumentError('Cannot parse connectToken');
+        }
+
+        if (!parameters || !parameters.CallSid || !signalingReconnectToken) {
+          throw new InvalidArgumentError('Invalid connectToken');
+        }
       }
-    }
 
-    let isReconnect = false;
-    let twimlParams: Record<string, string> = {};
-    const callOptions: Call.Options = {
-      enableImprovedSignalingErrorPrecision:
+      let isReconnect = false;
+      let twimlParams: Record<string, string> = {};
+      const callOptions: Call.Options = {
+        enableImprovedSignalingErrorPrecision:
         !!this._options.enableImprovedSignalingErrorPrecision,
-      rtcConfiguration: options.rtcConfiguration,
-      voiceEventSidGenerator: this._options.voiceEventSidGenerator,
-    };
+        rtcConfiguration: options.rtcConfiguration,
+        voiceEventSidGenerator: this._options.voiceEventSidGenerator,
+      };
+
+      if (signalingReconnectToken && parameters) {
+        isReconnect = true;
+        callOptions.callParameters = parameters;
+        callOptions.reconnectCallSid = parameters.CallSid;
+        callOptions.reconnectToken = signalingReconnectToken;
+        twimlParams = customParameters || twimlParams;
+      } else {
+        twimlParams = options.params || twimlParams;
+      }
 
-    if (signalingReconnectToken && parameters) {
-      isReconnect = true;
-      callOptions.callParameters = parameters;
-      callOptions.reconnectCallSid = parameters.CallSid;
-      callOptions.reconnectToken = signalingReconnectToken;
-      twimlParams = customParameters || twimlParams;
-    } else {
-      twimlParams = options.params || twimlParams;
-    }
+      const activeCall = this._activeCall = await this._makeCall(twimlParams, callOptions, isReconnect);
 
-    const activeCall = this._activeCall = await this._makeCall(twimlParams, callOptions, isReconnect);
+      // Make sure any incoming calls are ignored
+      this._calls.splice(0).forEach(call => call.ignore());
 
-    // Make sure any incoming calls are ignored
-    this._calls.splice(0).forEach(call => call.ignore());
+      // Stop the incoming sound if it's playing
+      this._soundcache.get(Device.SoundName.Incoming).stop();
 
-    // Stop the incoming sound if it's playing
-    this._soundcache.get(Device.SoundName.Incoming).stop();
+      activeCall.accept({ rtcConstraints: options.rtcConstraints });
+      this._publishNetworkChange();
+      return activeCall;
+    };
 
-    activeCall.accept({ rtcConstraints: options.rtcConstraints });
-    this._publishNetworkChange();
-    return activeCall;
+    return this._connectPromise = _connect().finally(() => {
+      this._connectPromise = null;
+    });
```

Comment:
Instead of wrapping the whole connect method, we should only wrap the async calls that are under this method. Good thing is, there's only one - makeCall. See the `await _makeCall` here. Maybe save the promise that it returns and use that to 'pause` the setInputDevice call.